Add ERC: Named NFT (#320)
ethereum/ERCs: ERCS/erc-7653.md

---
eip: 7653
title: Named NFT

> description: An extension to ERC-721 based on ERC-7632.

Reviewer: Your description doesn't really say anything about what's in this proposal. It only says where to look for more information.

It'd be nice to see at a glance what's actually in this document.

author: Zainan Victor Zhou (@xinbenlv)

> discussions-to: https://ethereum-magicians.org/t/erc-tbd-named-nfts-extending-erc-721/18550

Reviewer 2: ```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-7653-named-nfts-erc-721-extension/18550
```

status: Draft
type: Standards Track
category: ERC
created: 2024-03-13
requires: 165, 721, 7632
---

## Abstract

> An extension to `ERC-721` based on `ERC-7632` to enable named `ERC-721`

Reviewer: ```suggestion
An extension to [ERC-721](./eip-721.md) based on [ERC-7632](./eip-7632.md) to enable named ERC-721.
```

Don't use backticks to bypass the linter.

> An extension to `ERC-721` based on `ERC-7632` to enable named `ERC-721`

Reviewer: Your abstract is certainly terse, but it doesn't contain enough information about what the proposal actually does. What is a named NFT? How do the names work?

> An extension to [ERC-721](./eip-721.md) based on [ERC-7632](./eip-7632.md) to enable named ERC-721.

Reviewer: How about:

> An extension to [ERC-721](./eip-721.md) based on [ERC-7632](./eip-7632.md) to enable named ERC-721. A named NFT is a token given a unique (per issuing contract) string name, that is usually human readable, as an identifier. These named NFTs can be manipulated (eg. transferred, approved, etc.) both by their token ID and by their name.

## Motivation

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

> 1. A compliant contract MUST implement `ERC-721` and following interface

Reviewer: ```suggestion
1. A compliant contract MUST implement `ERC721` and following interface
```

    ///  not the current owner. Throws if `_to` is the zero address. Throws if
    ///  `_tokenId` is not a valid NFT.
    /// @param _from The current owner of the NFT
    /// @param _to The new owner
    /// @param _tokenId The NFT to transfer
    function transferFromByName(address _from, address _to, string memory _tokenName) external payable;

    /// @notice Change or reaffirm the approved address for an NFT
    /// @dev The zero address indicates there is no approved address.
    ///  Throws unless `msg.sender` is the current NFT owner, or an authorized
    ///  operator of the current owner.
    /// @param _approved The new approved NFT controller
    /// @param _tokenId The NFT to approve
    function approveByName(address _approved, string memory _tokenName) external payable;

    /// @notice Get the approved address for a single NFT
    /// @dev Throws if `_tokenId` is not a valid NFT.
    /// @param _tokenId The NFT to find the approved address for
    /// @return The approved address for this NFT, or the zero address if there is none
    function getApprovedByName(string memory _tokenName) external view returns (address);
}
```

> 2. A compliant contract MUST implement the `ERC-165`

Reviewer: ```suggestion
2. A compliant contract MUST implement the `ERC165` interface
```

<!-- TODO: inlcude `ERC-165` interface id  -->

## Rationale

> Needs discussion

Reviewer: ```suggestion
Needs discussion <!-- TODO -->
```

## Backwards Compatibility

> This ERC is designed to be backward compatible with `ERC-721`

Reviewer: ```suggestion
This ERC is designed to be backward compatible with ERC-721.
```

## Security Considerations

> Needs discussion.

Reviewer: ```suggestion
Needs discussion. <!-- TODO -->
```
